pennylane/resource/specs.py

```diff
@@ -117,31 +135,244 @@ def _specs_qjit(qjit, level, compute_depth, *args, **kwargs) -> CircuitSpecs:  #
         if os.path.exists(_RESOURCE_TRACKING_FILEPATH):
             os.remove(_RESOURCE_TRACKING_FILEPATH)
 
+
+def _preprocess_level_input(level, marker_to_level) -> list[int]:
+    """Preprocesses the level input to always return a sorted list of integers.
+
+    Args:
+        level (str | int | slice | iter[int | str]): The level input to preprocess
+        marker_to_level (dict[str, int]): Mapping from marker names to their associated level numbers
+    Returns:
+        list[int]: The preprocessed level input
+    """
+
+    if isinstance(level, (int, str)):
+        level = [level]
+    else:
+        level = list(level)
+
+    # Convert marker names to the associated level number
+    for i, lvl in enumerate(level):
+        if isinstance(lvl, str):
+            if lvl not in marker_to_level:
+                raise ValueError(f"Marker name '{lvl}' not found in the transform program.")
+            level[i] = marker_to_level[lvl]
+        elif isinstance(lvl, int):
+            if lvl < 0:
+                raise ValueError(
+                    "The 'level' argument to qml.specs for QJIT'd QNodes must be non-negative, "
+                    f"got {lvl}."
+                )
+
+    level_sorted = sorted(level)
+    if level != level_sorted:
+        warnings.warn(
+            "The 'level' argument to qml.specs for QJIT'd QNodes has been sorted to be in ascending order.",
+            UserWarning,
+        )
+
+    return level_sorted
+
+
+def _specs_qjit_intermediate_passes(
+    qjit, original_qnode, level, *args, **kwargs
+) -> (
+    SpecsResources | list[SpecsResources] | dict[str, SpecsResources | list[SpecsResources]]
+):  # pragma: no cover
+    # pylint: disable=import-outside-toplevel,too-many-branches,too-many-statements
+    from catalyst.python_interface.inspection import mlir_specs
+
+    # Note that this only gets transforms manually applied by the user
+    trans_prog = original_qnode.transform_program
+
+    single_level = isinstance(level, (int, str)) and not level in ("all", "all-mlir")
+
+    # Levels where qml.marker transforms are applied, needed since markers may be applied before or
+    # after the first MLIR transform, and ones after need to be incremented by 1 to account for the
+    # extra lowering pass
```

Comment:
Which lowering pass are you referring to?

Reply:
The JAXPR -> MLIR lowering. Basically we have 2 false layers that we insert into the user's transform program, level 0 "before any transforms" and level `n+1` (for `n` tape transforms) which is "after lowering to MLIR but before any MLIR passes"